staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/interface.go

```diff
@@ -0,0 +1,98 @@
+// clock, and any other test-facilitating infrastructure, to use in
+// each QueueSet is known to the factory, the client does not need to
+// pass this stuff in each call to NewQueueSet.
+type QueueSetFactory interface {
+	NewQueueSet(name string, concurrencyLimit, numQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) (QueueSet, error)
+}
+
+// QueueSet is the abstraction for the queuing and dispatching
+// functionality of one non-exempt priority level.  It covers the
+// functionality described in the "Assignment to a Queue", "Queuing",
+// and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type QueueSet interface {
+	// SetConfiguration updates the configuration
+	SetConfiguration(concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int, requestWaitLimit time.Duration) error
+
+	// Quiesce controls whether this system is quiescing.  Passing a
+	// non-nil handler means the system should become quiescent, a nil
+	// handler means the system should become non-quiescent.  A call
+	// to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.  If all the queues have no requests
+	// waiting nor executing while the system is quiescent then the
+	// handler will eventually be called with no locks held (even if
+	// the system becomes non-quiescent between the triggering state
+	// and the required call).
+	//
+	// The filter uses this for a priority level that has become
+	// undesired, setting a handler that will cause the priority level
+	// to eventually be removed from the filter if the filter still
+	// wants that.  If the filter later changes its mind and wants to
+	// preserve the priority level then the filter can use this to
+	// cancel the handler registration.
+	Quiesce(EmptyHandler)
+
+	// Wait, in the happy case, shuffle shards the given request into
+	// a queue and eventually dispatches the request from that queue.
+	// Dispatching means to return with `quiescent==false` and
+	// `execute==true`.  In one unhappy case the request is
+	// immediately rebuffed with `quiescent==true` (which tells the
+	// filter that there has been a timing splinter and the filter
+	// re-calcuates the priority level to use); in all other cases
+	// `quiescent` will be returned `false` (even if the system is
+	// quiescent by then).  In the non-quiescent unhappy cases the
+	// request is eventually rejected, which means to return with
+	// `execute=false`.  In the happy case the caller is required to
+	// invoke the returned `afterExecution` after the request is done
+	// executing.  The hash value and hand size are used to do the
+	// shuffle sharding.
+	Wait(hashValue uint64) (quiescent, execute bool, afterExecution func())
+}
+
+// EmptyHandler can be notified that something is empty
```

Comment:
The comment should explain why this is useful?

Reply:
Apparently the design rationale needs better commentary.  I haven't decided yet where is the best place to write it, because the design takes two layers into account.

Reply:
This is a comment problem, not a design problem; specifically "something" in the sentence makes it tautological and therefore unhelpful.

An example useful comment: `// EmptyHandler is used to notify callers when all the queues of a QueueSet have been drained.`

Reply:
Updated the explanation:
```
// EmptyHandler can be notified that something is empty.  This is used
// for a final cleanup action when a QueueSet is quiescing and
// no longer has queued or executing requests.
```

EmptyHandler is used here:
https://github.com/kubernetes/kubernetes/pull/84544/files#diff-253f1af862522d6e14b87c7ba89323b4R580-R587

Reply:
Updated to use the suggested:
```
// EmptyHandler is used to notify callers when all the queues
// of a QueueSet have been drained.
```